pkg/service/running_backups_registry.go

```diff
@@ -0,0 +1,161 @@
+package service
+
+import (
+	"context"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/aerospike/aerospike-backup-service/v3/pkg/model"
+	"github.com/aerospike/aerospike-backup-service/v3/pkg/util"
+)
+
+// RunningBackupsRegistry defines the interface for managing running backups and their statuses.
+type RunningBackupsRegistry interface {
+	// register add a new backup handler for a specific routine and job type.
+	register(string, jobType, CancelableBackupHandler)
+	// finishWithError remove backup from the registry.
+	finishWithError(routineName string, job jobType)
+	// FinishFull remove backup from registry and update last success timestamp.
+	FinishFull(routineName string, time time.Time)
+	// FinishIncremental remove incremental backup from registry and update last success timestamp.
+	FinishIncremental(routineName string, time time.Time)
+	// CurrentStat get the current backup statistics for a routine.
+	CurrentStat(string) *model.CurrentBackups
+	// GetAllCurrentStats all current backups statistics.
+	GetAllCurrentStats() map[string]*model.CurrentBackups
+	// Cancel all ongoing backups for a specific routine.
+	Cancel(string)
+}
+type registryKey struct {
+	routineName string
+	job         jobType
+}
+
+// RunningBackupsRegistryImpl implements the RunningBackupsRegistry interface.
+type RunningBackupsRegistryImpl struct {
+	handlers       *util.SafeMap[registryKey, CancelableBackupHandler]
+	lastSuccessful *util.SafeMap[string, *model.LastBackupRun]
+}
+
+var _ = (*RunningBackupsRegistryImpl)(nil)
+
+// NewRunningBackupsRegistry creates a new instance of RunningBackupsRegistryImpl.
+func NewRunningBackupsRegistry() *RunningBackupsRegistryImpl {
+	return &RunningBackupsRegistryImpl{
+		handlers:       util.NewSafeMap[registryKey, CancelableBackupHandler](),
+		lastSuccessful: util.NewSafeMap[string, *model.LastBackupRun](),
+	}
+}
+
+// SyncBackupHistoryFromStorage updates the backup registry with the most recent backup timestamps
+// found in the storage backends. It scans all backup routines in parallel.
+func SyncBackupHistoryFromStorage(
```

Comment:
This function is called at service startup only to collect the last execution time for routines. It is better not to block here until the data is collected.